daemon/src/wallet.rs

```diff
@@ -354,17 +356,70 @@ where
     }
 }
 
+struct LockedUtxos {
+    inner: HashSet<(Instant, OutPoint)>,
+    time_to_lock: Duration,
+}
+
+impl LockedUtxos {
+    fn new(network: Network) -> Self {
+        let time_to_lock = match network {
+            Network::Bitcoin => Duration::from_secs(20 * 60),
+            Network::Testnet | Network::Signet | Network::Regtest => Duration::from_secs(2 * 60),
+        };
+
+        Self {
+            inner: HashSet::default(),
+            time_to_lock,
+        }
+    }
+
+    /// Add new elements to the set of locked UTXOs.
+    fn extend<T: IntoIterator<Item = OutPoint>>(&mut self, utxos: T) {
+        let now = Instant::now();
+        let utxos = utxos.into_iter().map(|utxo| (now, utxo));
+
+        self.inner.extend(utxos);
+    }
+
+    /// Return the list of locked UTXOs.
+    ///
+    /// Before creating the list, it removes all elements which should
+    /// no longer be part of the set of locked UTXOs.
+    fn list(&mut self) -> Vec<OutPoint> {
+        self.remove_expired();
```

Comment:
❓ This removes the UTXOs that were actually spent (i.e. where a tx was published that spends them)?

Reply:
Not quite. The code doesn't actually reason about whether a UTXO was spent or not. 

UTXOs are locked from the perspective of our wallet up to a time based on a configurable value `time_to_lock`. This `LockedUtxos` data structure will eventually unlock them if enough time has passed. Some of those locked UTXOs may have been used in transactions, but enough time should have passed that the wallet doesn't attempt to use them again (and if the wallet does it would just fail). Others were locked but never used because contract setup failed, so it is desirable to unlock them (this is the feature we're adding with this PR).

Reply:
We may want to couple this unlock time with our sync interval in some form.

Reply:
Can you describe what you're suggesting in more detail, @thomaseizinger?

Reply:
Whether or not it is safe to discard locked UTXOs depends on our sync picking up the new transactions (and thus marking it internally as spent. `LockedUtxos` only exists because we need to reliably bridge the time gap between signing a transaction and the UTXOs showing up during the sync.

If we set `time_to_lock` to something like 4 * sync_internal, we should be reasonable safe.

Reply:
Note that we do pick up UTXOs used by transactions in the mempool so we don't actually need to wait for a confirmation.

Reply:
Thanks for the explanation, that makes sense to me now.

Basically, by setting the `time_to_lock` field to `4 * sync_interval` we're saying that we don't expect any locked UTXOs to end up being used if they were not picked up by the  wallet after 4 `sync_interval`s.

Reply:
> Basically, by setting the `time_to_lock` field to `4 * sync_interval` we're saying that we don't expect any locked UTXOs to end up being used if they were not picked up by the wallet after 4 `sync_interval`s.

Correct.